cli/cmd/inject.go

```diff
@@ -209,105 +214,148 @@ func InjectYAML(in io.Reader, out io.Writer, version string) error {
+	// CONDUIT_PROXY_CONTROL_URL variable must be set to localhost for the following reasons:
+	//	1. According to https://github.com/kubernetes/minikube/issues/1568, minikube has an issue
+	//     where pods are unable to connect to themselves through their associated service IP.
+	//     Setting the CONDUIT_PROXY_CONTROL_URL to localhost allows the proxy to bypass kube DNS
+	//     name resolution as a workaround to this issue.
+	//  2. We avoid the TLS overhead in encrypting and decrypting intra-pod traffic i.e. traffic
+	//     between containers in the same pod.
+	//  3. Using a Service IP instead of localhost would mean intra-pod traffic would be load-balanced
+	//     across all controller pod replicas. This is undesirable as we would want all traffic between
+	//	   containers to be self contained.
+	//  4. We skip recording telemetry for intra-pod traffic within the control plane.
+	switch meta.Kind {
+	case "Deployment":
+		var deployment v1beta1.Deployment
+		if err := yaml.Unmarshal(bytes, &deployment); err != nil {
+			return nil, err
+		}
+
+		if deployment.Name == ControlPlanePodName && deployment.Namespace == controlPlaneNamespace {
+			DNSNameOverride = LocalhostDNSNameOverride
+		}
+
+		obj = &deployment
+		k8sLabels[k8s.ProxyDeploymentLabel] = deployment.Name
+		podTemplateSpec = &deployment.Spec.Template
+	case "ReplicationController":
+		var rc v1.ReplicationController
+		if err := yaml.Unmarshal(bytes, &rc); err != nil {
+			return nil, err
+		}
+
+		obj = &rc
+		k8sLabels[k8s.ProxyReplicationControllerLabel] = rc.Name
+		podTemplateSpec = rc.Spec.Template
+	case "ReplicaSet":
+		var rs v1beta1.ReplicaSet
+		if err := yaml.Unmarshal(bytes, &rs); err != nil {
+			return nil, err
+		}
+
+		obj = &rs
+		k8sLabels[k8s.ProxyReplicaSetLabel] = rs.Name
+		podTemplateSpec = &rs.Spec.Template
+	case "Job":
+		var job batchV1.Job
+		if err := yaml.Unmarshal(bytes, &job); err != nil {
+			return nil, err
+		}
+
+		obj = &job
+		k8sLabels[k8s.ProxyJobLabel] = job.Name
+		podTemplateSpec = &job.Spec.Template
+	case "DaemonSet":
+		var ds v1beta1.DaemonSet
+		if err := yaml.Unmarshal(bytes, &ds); err != nil {
+			return nil, err
+		}
+
+		obj = &ds
+		k8sLabels[k8s.ProxyDaemonSetLabel] = ds.Name
+		podTemplateSpec = &ds.Spec.Template
+	}
+
+	// If we don't inject anything into the pod template then output the
+	// original serialization of the original object. Otherwise, output the
+	// serialization of the modified object.
+	output := bytes
+	if podTemplateSpec != nil && injectPodTemplateSpec(podTemplateSpec, DNSNameOverride, version, k8sLabels) {
+		var err error
+		output, err = yaml.Marshal(obj)
+		if err != nil {
+			return nil, err
+		}
+		// Lists are a little different than the other types. There's no immediate
+		// pod template. Because of this, we do a recursive call for each element
+		// in the list (instead of just marshaling the injected pod template).`
+	} else if meta.Kind == "List" {
```

Comment:
That's what the comment above is about. the switch assumes that there will be a single pod template that gets injected. Because lists have no pod template, they need to get handled differently.

Reply:
(The whole thing should probably be refactored eventually, the tooling that kubectl uses is exported and available. That's way too big of a change here and I'm not sure it is worth the effort yet)

Reply:
i was pointing out that checking for `meta.Kind == "List"` is logically more in-line with `switch meta.Kind`, rather than as an else clause on `podTemplateSpec != nil && injectPodTemplateSpec(...)`. it accomplishes the same thing, but i think is a bit easier to follow:

```diff
$ git diff
diff --git a/cli/cmd/inject.go b/cli/cmd/inject.go
index 483c21d..78f2e29 100644
--- a/cli/cmd/inject.go
+++ b/cli/cmd/inject.go
@@ -342,6 +342,11 @@ func injectResource(bytes []byte, version string) ([]byte, error) {
                obj = &ds
                k8sLabels[k8s.ProxyDaemonSetLabel] = ds.Name
                podTemplateSpec = &ds.Spec.Template
+       case "List":
+               // Lists are a little different than the other types. There's no immediate
+               // pod template. Because of this, we do a recursive call for each element
+               // in the list (instead of just marshaling the injected pod template).
+               return injectList(bytes, version)
        }

        // If we don't inject anything into the pod template then output the
@@ -354,11 +359,6 @@ func injectResource(bytes []byte, version string) ([]byte, error) {
                if err != nil {
                        return nil, err
                }
-               // Lists are a little different than the other types. There's no immediate
-               // pod template. Because of this, we do a recursive call for each element
-               // in the list (instead of just marshaling the injected pod template).
-       } else if meta.Kind == "List" {
-               return injectList(bytes, version)
        }

        return output, nil
```